File: Sorting/QuickSort.java
public class QuickSort {

    public static void quickSort(int[] arr) {
        quickSort(arr, 0, arr.length-1);
    }

    public static void quickSort(int[] arr, int leftIndex, int rightIndex) {
        int arrLength = rightIndex + 1 - leftIndex;

        if (arrLength > 1) {
            int partitionIndex = partition(arr, leftIndex, rightIndex);

            quickSort(arr, leftIndex, partitionIndex-1);
            quickSort(arr, partitionIndex+1, rightIndex);
        }
    }

    public static int partition(int[] arr, int leftIndex, int rightIndex) {
        int pivot = arr[rightIndex];
        int partitionIndex = leftIndex;

        for (int i = leftIndex; i < rightIndex; i++) {
            if (arr[i] < pivot) {
                swap(arr, i, partitionIndex);
                partitionIndex++;
            }
        }
        swap(arr, partitionIndex, rightIndex);

        return partitionIndex;
    }

    public static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: Sorting/quickSort.js
const swap = (items, leftIndex, rightIndex) => {
  const temp = items[leftIndex]
  items[leftIndex] = items[rightIndex]
  items[rightIndex] = temp
}

const partition = (items, left, right) => {
  let pivot = items[Math.floor((right + left) / 2)]
  let i = left
  let j = right
  while (i <= j) {
    while (items[i] < pivot) {
      i++
    }
    while (items[j] > pivot) {
      j--
    }
    if (i <= j) {
      swap(items, i, j)
      i++
      j--
    }
  }
  return i
}

const quickSort = (items, left, right) => {
  let index
  if (items.length > 1) {
    index = partition(items, left, right)
    if (left < index - 1) {
      quickSort(items, left, index - 1)
    }
    if (index < right) {
      quickSort(items, index, right)
    }
  }
  return items
}

export default quickSort
